Convert to use signal input and output

pankajparkar opened the pull request:
# Convert to use signal `input` and `output`
This update will make the components use the latest signal inputs, outputs, computed, effect.
Library code will be following the reactive style.

Changes include
- Convert `@Input` to `input` signal
- Convert `@Ouput` to `output` signal
- Use `setInput` on component instance
- Replaced `ngOnChanges`, with `computed` / `efffect`
- Used `ChangeDetectionStrategy.OnPush` strategy
- Updates affected tests to use signal
- Update tests with signal -✅

devoto13 reviewed (commented):
Thanks for the PR! Looks very nice 💪

Left couple small comments. Please let me know if you want to address them or plan to push some more changes, otherwise I can address myself and land the PR.

pankajparkar commented:
Yes, few tests are failing will fix them also along with your comments.

Thanks for having a initial look, cheers. Thanks

pankajparkar commented on src/lib/fontawesome.module.ts:
```diff
@@ -7,6 +7,11 @@ import { FaLayersComponent } from './layers/layers.component';
 import { FaStackComponent } from './stack/stack.component';
 
+/**
+ * @deprecated
```
Make sense, reverted deprecated comment

https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/a5980d729bfa8cbfd5ba25c6fa4ae6daf187d4c6

pankajparkar commented on src/lib/icon/icon.component.spec.ts:
```diff
@@ -87,7 +88,7 @@ describe('FaIconComponent', () => {
     expect(queryByCss(fixture, 'svg').classList.contains('fa-spin')).toBeFalsy();
 
-    fixture.componentInstance.iconComponent.animation = 'spin';
+    signalSetFn(fixture.componentInstance.iconComponent.animation[SIGNAL], 'spin');
```
@devoto13 

Oops, I never thought of that from an implementation perspective. I dig a bit and found it's a feature that we documented already. We allow consumers to update the `input` values programatically

Okay, coming back to it, why it has to be that way.

By design `input` signals are `readonly`, they can not be set from from component instance.
TBH, I'm not sure whether it's a good idea to manipulate the inputs of a component from parent/container component (against the idea of unidirectional data flow).

Anyways coming to solutioning part
## The possible approach to keep current feature intact is

[Convert all FaIconComponent `input` to `model`](https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/1fc72ff6d0a03ebaeeef11d545059676744f99de), so that we can call `iconComponent.animation.set('spin')` - Originally `model` signal usecase is bit different, but we would use it here to keep existing capability intact  

The new approaches (**BREAKING CHANGE**s)
1. `signalSetFn(fixture.componentInstance.iconComponent.animation[SIGNAL], 'spin');` (keeping current approach intact)
2. `faIconComponentRef.setInput('animation', this.animation())`

yaroslav-softwerk replied:
(1) looks good, let's use this solution.
(2) I don't think we can recommend it to the users as `signalSetFn` is not part of the Angular's public API and can change in the future.  
(3) is an option, but the drawback is that it does not offer any type-checking.

While I agree with you that changing inputs programmatically is a dubious practice, this was a common requests from users, so it's important to keep the capability.

Otherwise, I have checked the PR and everything looks good besides couple of small comments. Thanks again for working on it!

pankajparkar replied:
Cool - Updated code to use `model` in `FaIconComponent`
https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/2f0cd97ad251cb8ae91cddbdc23c208f8be62cda#diff-172cc29c974c733b93f2b619cbdc6bab9d124a261e82bee68a12c99860643657

pankajparkar commented on src/lib/icon/duotone-icon.component.spec.ts:
```diff
@@ -125,7 +125,7 @@ describe('FaDuotoneIconComponent', () => {
       createIcon() {
         const componentRef = this.container.createComponent(FaDuotoneIconComponent);
-        componentRef.instance.icon = faDummy;
+        componentRef.setInput('icon', faDummy);
```
Updated - https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/a8842d7d478f9533fbe4a3f0e61c2442c5e9baa2

But I will still update the code based on what decision we make on [other concern](https://github.com/FortAwesome/angular-fontawesome/pull/464#discussion_r1997654227)

pankajparkar replied:
This is addressed we're settled with `model`.

pankajparkar commented:
Only one unit test is failing so far (working on fixing it). Also, I will add more coverage with signal example to gain more confidence over the changes.

pankajparkar commented on docs/usage/features.md:
````diff
@@ -230,3 +232,55 @@ class HostComponent {
 }
 ```
+
+#### v1.1.0 Onwards
````
Updated - https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/ec8e3f7a6070506e25bc39a4ff276ddbbfa6cef0#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8

pankajparkar commented on docs/usage/features.md:
```diff
@@ -230,3 +232,55 @@ class HostComponent {
+
+  spinIcon() {
+    const iconComponent = this.iconComponent();
+    if (!container) {
```
Updated - https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/e420ca8d0a3fbea38f376c6c17927b5e55c9174f

pankajparkar commented on docs/usage/features.md:
```diff
@@ -230,3 +232,55 @@ class HostComponent {
+})
+class HostComponent {
+  readonly faUser = faUser;
+  readonly iconComponent = viewChild(FaIconComponent, { static: true });
```
Updated - https://github.com/FortAwesome/angular-fontawesome/pull/464/commits/e420ca8d0a3fbea38f376c6c17927b5e55c9174f

pankajparkar commented:
PR is ready for a review 

- Updated the PR description to reflect changes
- Updated readme.md
- Updated migration.md
- Updated changelog.md
- Updated release version

Let me know if I have missed anything else

devoto13 reviewed (approved):
Looks great! Thanks a lot for working on it 👍 

I just made one small change as I would prefer to release it together with the next major release of Angular.
